solarwinds_apm/oboe/sampler.py

```diff
@@ -0,0 +1,309 @@
+    warning = unparsed.get("warning")
+    return (
+        Settings(
+            sample_source=SampleSource.REMOTE,
+            sample_rate=sample_rate,
+            flags=flags,
+            timestamp=timestamp,
+            ttl=ttl,
+            buckets=buckets,
+            signature_key=signature_key,
+        ),
+        warning,
+    )
+
+
+class Sampler(OboeSampler):
+    def __init__(
+        self,
+        meter_provider: MeterProvider,
+        config: Configuration,
+        logger: Logger,
+        initial: Any,
+    ):
+        super().__init__(meter_provider=meter_provider, logger=logger)
+        if config.tracing_mode is not None:
+            self._tracing_mode = (
+                TracingMode.ALWAYS
+                if config.tracing_mode
+                else TracingMode.NEVER
+            )
+        else:
+            self._tracing_mode = None
+        self._trigger_mode = config.trigger_trace_enabled
+        self._transaction_settings = config.transaction_settings
+        self._ready = threading.Event()
+        if initial:
+            self.update_settings(initial)
+
+    def __str__(self) -> str:
+        return f"Sampler{self._tracing_mode}({self._trigger_mode}) {super().__str__(self)}"
+
+    @property
+    def tracing_mode(self):
+        return self._tracing_mode
+
+    @property
+    def trigger_mode(self):
+        return self._trigger_mode
+
+    @property
+    def transaction_settings(self):
+        return self._transaction_settings
+
+    def wait_until_ready(self, timeout: int) -> bool:
+        """
+        Waits until the sampler is ready.
+        """
+        return self._ready.wait(timeout)
+
+    @override
+    def local_settings(
+        self,
+        parent_context: "Context" | None,
+        trace_id: int,
+        name: str,
+        kind: SpanKind | None = None,
+        attributes: Attributes = None,
+        links: Sequence["Link"] | None = None,
+        trace_state: "TraceState" | None = None,
+    ) -> LocalSettings:
+        """
+        Returns local settings.
+        """
+        settings = LocalSettings(
+            tracing_mode=self.tracing_mode, trigger_mode=self.trigger_mode
+        )
+        if (
+            self.transaction_settings is None
+            or len(self.transaction_settings) == 0
+        ):
+            return settings
+        meta = http_span_metadata(kind, attributes)
+        identifier = (
+            meta["url"] if meta["http"] else f"{SpanKind(kind).name}:{name}"
+        )
+        for transaction_setting in self.transaction_settings:
+            if transaction_setting.matcher and transaction_setting.matcher(
+                identifier
+            ):
+                settings.tracing_mode = (
+                    TracingMode.ALWAYS
+                    if transaction_setting.tracing
+                    else TracingMode.NEVER
+                )
+                break
+        return settings
+
+    @override
+    def request_headers(
```

Comment:
Similar to [below comment](https://github.com/solarwinds/apm-python/pull/534/files#r1992145178): Which trigger trace scenarios were tested with HttpSampler? Does this co-operate with current custom Propagator [inject](https://github.com/solarwinds/apm-python/blob/4caf5852935f3b32bdccf0bf1206f5132bd16291/solarwinds_apm/propagator.py#L43-L51) and [extract](https://github.com/solarwinds/apm-python/blob/4caf5852935f3b32bdccf0bf1206f5132bd16291/solarwinds_apm/propagator.py#L43-L51)?

Reply:
The trigger trace scenarios are in test_oboe_sampler.py file. 
HttpSampler and JsonSampler are the wrapper for OboeSampler (which does most of the sampling decision logic)

Reply:
For the cooperation with Propagator, I have to check. 
In this PR, it tries to get the TraceOptions from `parent_context`. 
https://github.com/solarwinds/apm-python/blob/NH-102108/solarwinds_apm/oboe/sampler.py#L241-L263
and set the response to the `parent_context`
https://github.com/solarwinds/apm-python/blob/NH-102108/solarwinds_apm/oboe/sampler.py#L266-L291

Reply:
Resolving, also based on https://github.com/solarwinds/apm-python/pull/534#issuecomment-2734886061